### PR TITLE
HEEDLS-800 add emails

### PR DESCRIPTION
**Pull request by JBsoftwire**
### JIRA link
_[HEEDLS-800](https://softwiretech.atlassian.net/browse/HEEDLS-800)_

### Description
Added email titles to cards that did not already have emails, made candidate number title elements hidden (so still searchable), and made the means of inclusion of emails in titles consistent (there were two methods in use before).

### Screenshots
image
image
image
image
image

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [X] Run the formatter and made sure there are no IDE errors.
- [x] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [X] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [ ] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [X] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/AllDelegates/_SearchableDelegateCard.cshtml [hunk 1]
```diff
@@ -7,7 +7,8 @@
   <details class="nhsuk-details nhsuk-expander">
     <summary class="nhsuk-details__summary">
       <span class="nhsuk-details__summary-text" id="@Model.DelegateInfo.Id-name">
-        <span class="searchable-element-title" name="name">@Model.DelegateInfo.TitleName</span> <span class="searchable-element-candidate-number">(@Model.DelegateInfo.CandidateNumber)</span>
+        <span class="searchable-element-title" name="name">@Model.DelegateInfo.TitleName</span>
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/AllDelegates/_SearchableDelegateCard.cshtml:10]: The change includes the email in the `searchable-element-title` span, so the JS search and sort will start returning results based on the email, but the non-JS search won't take it into account. From reading the ticket, it feels to me like we just want to change the display without changing the search functionality. Thoughts on this?

Same applies to the `_AddGroupDelegateCard`

**Review by DanBloxham-sw — CHANGES_REQUESTED** [review]
These changes have introduced an inconsistency in the search and sort on these pages that will need to be dealt with, though I'm not 100% which way we want to go with it.

File: DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/AllDelegates/_SearchableDelegateCard.cshtml [hunk 2]
```diff
@@ -7,7 +7,8 @@
   <details class="nhsuk-details nhsuk-expander">
     <summary class="nhsuk-details__summary">
       <span class="nhsuk-details__summary-text" id="@Model.DelegateInfo.Id-name">
-        <span class="searchable-element-title" name="name">@Model.DelegateInfo.TitleName</span> <span class="searchable-element-candidate-number">(@Model.DelegateInfo.CandidateNumber)</span>
+        <span class="searchable-element-title" name="name">@Model.DelegateInfo.TitleName</span>
+        <span class="searchable-element-candidate-number nhsuk-u-visually-hidden">@Model.DelegateInfo.CandidateNumber</span>
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/AllDelegates/_SearchableDelegateCard.cshtml:11]: I agree with Dan that the change as is causes the search functionality between JS and no JS to diverge. I also think that this visually hidden span can probably be removed in favour of adding the class to the Candidate number display within the card.

This should probably be checked with HEE whether they want searching on the email or not.

**Review by AlexJacksonDS — CHANGES_REQUESTED** [review]
I agree with Dan + some extra tidying that I think could happen (assuming it doesn't break search functionality)

**Comment by JBsoftwire** [comment]
I'm pushing this so people can look at it, but I haven't been able to test the JS search due to problems with `npm run build`. I'll fix that, test this and update as soon as I can.

**Comment by JBsoftwire** [comment]
There are some UI test errors, I'm looking into them (apparently these don't fail checks).

**Comment by JBsoftwire** [comment]
UI errors fixed with a new partial (they also revealed a JS search bug).

File: DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/CourseDelegates/_SearchableCourseDelegateCard.cshtml [hunk 3]
```diff
@@ -18,7 +20,7 @@
           <dt class="nhsuk-summary-list__key">
             Delegate ID
           </dt>
-          <dd class="nhsuk-summary-list__value" data-name-for-sorting="delegate-id">
+          <dd class="nhsuk-summary-list__value searchable-element-candidate_number" data-name-for-sorting="delegate-id">
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/CourseDelegates/_SearchableCourseDelegateCard.cshtml:23]: candidate_number? Should this be candidate-number?

File: DigitalLearningSolutions.Web/Helpers/DisplayStringHelper.cs [hunk 4]
```diff
@@ -59,6 +59,11 @@ public static string GetNameWithEmailForDisplay(string name, string? email)
                 : null;
         }
 
+        public static string GetEmailDisplayString(string? email)
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Web/Helpers/DisplayStringHelper.cs:62]: Could probably do with a couple of simple unit tests for this

**Review by AlexJacksonDS — CHANGES_REQUESTED** [review]
Looks good, just a couple of tiny things to resolve

**Pull request merged**